### PR TITLE
[WIP] Move expensive properties to "BrowseObject" rules

### LINKED ISSUE #2966: PERF: Clean up unused properties in file rules
File rules are used to produce the project tree through dataflow.  Currently, the project system often reuse browse object rules with file rules, which adds extra properties, which are evaluated, but never used by any feature.  This adds overhead to CPU/memory usages.

We want to get rid of those properties in file context rules (Compile/None/Content/AdditionalFiles/EmbeddedResource)

Identity FileNameAndExtension URL Extension.

The impact is bigger for rules covering lots of files in the project, so focus on those rules first.

### PR DESCRIPTION
**Pull request by Pilchie**
Fixes #2966 (or at least improves it).

NOTE: I'm currently unable to test this, because I can't create projects in the hive with project-system bits, regardless of whether my changes are included or not.
<details>**Customer scenario**

What does the customer do to get into this situation, and why do we think this
is common enough to address in Escrow.  (Granted, sometimes this will be
obvious "Open project, VS crashes" but in general, I need to understand how
common a scenario is)

**Bugs this fixes:** 

(either VSO or GitHub links)

**Workarounds, if any**

Also, why we think they are insufficient for RC vs. RC2, RC3, or RTW

**Risk**

This is generally a measure our how central the affected code is to adjacent
scenarios and thus how likely your fix is to destabilize a broader area of code

**Performance impact**

(with a brief justification for that assessment (e.g. "Low perf impact because no extra allocations/no complexity changes" vs. "Low")

**Is this a regression from a previous update?**

**Root cause analysis:**

How did we miss it?  What tests are we adding to guard against it in the future?

**How was the bug found?**

(E.g. customer reported it vs. ad hoc testing)
</details>

### CODE REVIEW

**Comment by Pilchie** [comment]
@lifengl

**Comment by lifengl** [comment]
This change carries some risks.  The top issue is that the new F# provider now uses the FullPath property in the tree, so we cannot remove that property, until we have a better solution for them.

**Comment by Pilchie** [comment]
So @davkean / @lifengl thoughts on taking this for 15.7?

**Comment by Pilchie** [comment]
Also: I had a follow-up question - looking at `Compile.xaml` and `Compile.BrowseObject.xaml` some properties are duplicated between the two, but not all.  Sometimes they are different types (`StringProperty` versus `EnumProperty` for example).  What's the expectation there?

**Comment by Pilchie** [comment]
Finally - @lifengl what's the best way to measure the impact on memory usage from this change to get a sense of reward for the risk?

**Comment by lifengl** [comment]
We have to measure the impact.  I don’t think we should take it to 15.7.

Sent from my phone

On Apr 24, 2018, at 9:54 AM, Kevin Pilch <notifications@github.com<mailto:notifications@github.com>> wrote:


Finally - @lifengl<https://eur03.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Flifengl&data=02%7C01%7C%7Ceef293587ad14a1404e608d5aa040b60%7C84df9e7fe9f640afb435aaaaaaaaaaaa%7C1%7C0%7C636601856695351919&sdata=0Bd%2BxoNZ914khNL3P4ucYCH%2B1UAHXjK5m%2FAIaVxmBQQ%3D&reserved=0> what's the best way to measure the impact on memory usage from this change to get a sense of reward for the risk?

**Comment by davkean** [comment]
@Pilchie You need to make setup changes to setup\Microsoft.VisualStudio.ProjectSystem.Managed.CommonFiles\CommonFiles.swr.

**Comment by Pilchie** [comment]
@davkean do you want me to do the setup for this and merge, or do you want to take over?

**Comment by Pilchie** [comment]
Okay - added the swr files.  @jmarolf how do I test that I did the right thing?

**Comment by jmarolf** [comment]
@Pilchie a build should produce a VSIX with all of the files zipped up.

**Comment by Pilchie** [comment]
@davkean any objections?  Suggestions on scenarios to verify?  I'll look at properties window at least...

**Comment by davkean** [comment]
Sorry missed this. 

- Do compare Properties window against legacy.
- I don't see a change to CSharp.xaml (and VB & F# equivalent if they have), I'd expect that.
- Check up-to-date.

**Comment by Pilchie** [comment]
It seems that this change broke the ability to see analyzer diagnostics in the Dependencies Node for some reason.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed/Microsoft.VisualStudio.ProjectSystem.Managed.csproj [hunk 1]
```diff
@@ -302,7 +302,11 @@
       <Generator>MSBuild:GenerateRuleSourceFromXaml</Generator>
       <SubType>Designer</SubType>
     </XamlPropertyRule>
+    <XamlPropertyRuleNoCodeBehind Include="ProjectSystem\Rules\AdditionalFiles.BrowseObject.xaml" />
```
Comment (davkean) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed/Microsoft.VisualStudio.ProjectSystem.Managed.csproj]: This is missing splits to ApplicationDefinition, AdditionalFiles, Page, & Folder.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/Rules/AdditionalFiles.BrowseObject.xaml [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+      DisplayName="Full Path"
+      ReadOnly="true"
+      Category="Misc"
+      Description="Location of the file.">
+    <StringProperty.DataSource>
+      <DataSource Persistence="Intrinsic" ItemType="AdditionalFiles" PersistedName="FullPath" SourceOfDefaultValue="AfterContext" />
+    </StringProperty.DataSource>
+  </StringProperty>
+
+  <StringProperty
```
[the remark this answers is not in the document]
Comment (Pilchie) [reply src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/Rules/AdditionalFiles.BrowseObject.xaml]: Not sure - I just moved all of these over, based on @lifengl's suggestion.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/Rules/AdditionalFiles.BrowseObject.xaml [hunk 3]
```diff
@@ -0,0 +1,61 @@
+﻿<?xml version="1.0" encoding="utf-8"?>
+<!--Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.-->
+<Rule
+  Name="AdditionalFiles"
+  DisplayName="Additional File"
+  PageTemplate="generic"
+  Description="File Properties"
```
Comment (davkean) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/Rules/AdditionalFiles.BrowseObject.xaml]: Please add PropertyPagesHidden="true" to all new "BrowseObjects" rules you added.

**Comment by davkean** [comment]
Closed in lieu of https://github.com/dotnet/project-system/pull/4016.

**Pull request closed without merging**